### PR TITLE
Refactor User models to use STI

### PR DESCRIPTION
**Pull request by JamieCleare2525**
## Context

To make developers lives easier, convert User models to use STI.

## Changes proposed in this pull request

- User models now use STI.
- Add `:type` to `users` table.
- Remove `:service` and `:support_user` attributes from `users` table. 
- Remove `Persona` model.

## Guidance to review

Test as much functionality as possible.

## Link to Trello card

https://trello.com/c/IB7IpN2x/148-refactor-user-models-to-use-single-table-inheritance

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated or [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/services/user_invite_service.rb [hunk 1]
```diff
@@ -1,16 +1,16 @@
 class UserInviteService
   include ServicePattern
 
-  def initialize(user, organisation)
+  def initialize(user, organisation, service)
```
Comment (elceebee) [inline app/services/user_invite_service.rb:4]: Can we not still determine the service from the user? Possible a method on the child classes, eg:
```
class Placements::User < User
  def service
    :placements
  end
end
```

So when you are sending the emails, we don't have to think about the possibility of sending an email to a `Claims::User` with the service `placements`

**Review by elceebee — COMMENTED** [review]
Wow, massive effort! Nice work. I've tested locally and can't find any issues. I do have a few comments though.

File: app/models/claims/support_user.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+#  type       :string
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+# Indexes
+#
+#  index_users_on_type_and_email  (type,email) UNIQUE
+#
+class Claims::SupportUser < User
+  def is_support_user?
```
[the remark this answers is not in the document]
Comment (JamieCleare2525) [reply app/models/claims/support_user.rb]: resolved in [0b42724](https://github.com/DFE-Digital/itt-mentor-services/pull/168/commits/0b42724ac7040f437f56b077f6d0a7e70a85fea9)

File: app/models/dfe_sign_in_user.rb [hunk 3]
```diff
@@ -45,30 +45,10 @@ def self.load_from_session(session)
...
-      begin
-        user = user_klass.find_by(email:)
-
-        if service == :placements && user&.support_user?
-          user = user.becomes(Placements::SupportUser)
-        end
-
-        user
-      end
+    @user ||= User.find_by(email:)
```
[the remark this answers is not in the document]
Comment (JamieCleare2525) [reply app/models/dfe_sign_in_user.rb]: resolved in [0b42724](https://github.com/DFE-Digital/itt-mentor-services/pull/168/commits/0b42724ac7040f437f56b077f6d0a7e70a85fea9)

File: app/models/placements/support_user.rb [hunk 4]
```diff
@@ -2,19 +2,20 @@
...
 #
 # Indexes
 #
-#  index_users_on_service_and_email  (service,email) UNIQUE
+#  index_users_on_type_and_email  (type,email) UNIQUE
 #
-class Placements::SupportUser < Placements::User
-  default_scope { support_users }
+class Placements::SupportUser < User
+  def is_support_user?
```
[the remark this answers is not in the document]
Comment (JamieCleare2525) [reply app/models/placements/support_user.rb]: resolved in [0b42724](https://github.com/DFE-Digital/itt-mentor-services/pull/168/commits/0b42724ac7040f437f56b077f6d0a7e70a85fea9)

File: app/models/user.rb [hunk 5]
```diff
@@ -2,35 +2,35 @@
...
 
-  scope :support_users, -> { where(support_user: true) }
+  scope :claims, -> { where(type: "Claims::User") }
+  scope :placements, -> { where(type: "Placements::User") }
 
   def full_name
     "#{first_name} #{last_name}".strip
   end
+
+  def is_support_user?
```
[the remark this answers is not in the document]
Comment (JamieCleare2525) [reply app/models/user.rb]: resolved in [0b42724](https://github.com/DFE-Digital/itt-mentor-services/pull/168/commits/0b42724ac7040f437f56b077f6d0a7e70a85fea9)

File: app/services/user_invite_service.rb (hunk 1 above)
Comment (JamieCleare2525) [reply app/services/user_invite_service.rb]: resolved in [0b42724](https://github.com/DFE-Digital/itt-mentor-services/pull/168/commits/0b42724ac7040f437f56b077f6d0a7e70a85fea9)

**Review by elceebee — APPROVED** [review]
Thanks for making all these changes! It looks good to me.

**Review by CatalinVoineag — APPROVED** [review]
This looks good to me. Well done, great work!

**Pull request merged**